sys/riscv/pmap.c

```diff
@@ -84,27 +84,71 @@ inline pte_t pte_protect(pte_t pte, vm_prot_t prot) {
  * Physical map management.
  */
 
+static MTX_DEFINE(kmap_l0_lock, MTX_SPIN);
+static paddr_t kmap_l0_pages[Ln_ENTRIES];
+
 static void update_kernel_pd(pmap_t *umap) {
   pmap_t *kmap = pmap_kernel();
 
-  size_t halfpage = PAGESIZE / 2;
-  memcpy(phys_to_dmap(umap->pde) + halfpage, phys_to_dmap(kmap->pde) + halfpage,
-         halfpage);
+  assert(umap != kmap);
+
+  WITH_MTX_LOCK (&kmap_l0_lock) {
+    if (umap->md.generation == kmap->md.generation)
+      return;
+
+    size_t halfpage = PAGESIZE / 2;
+    void *new_kpd = phys_to_dmap(kmap->pde) + halfpage;
+    void *old_kpd = phys_to_dmap(umap->pde) + halfpage;
+    memcpy(old_kpd, new_kpd, halfpage);
 
-  umap->md.generation = kmap->md.generation;
+    umap->md.generation = kmap->md.generation;
+  }
 }
 
-void pmap_md_activate(pmap_t *umap) {
+void pmap_md_growkernel(vaddr_t old_kva, vaddr_t new_kva) {
+  /* Did we change top kernel PD? If so force updates to user pmaps! */
+  if ((old_kva & ~L0_OFFSET) == (new_kva & ~L0_OFFSET))
+    return;
+
   pmap_t *kmap = pmap_kernel();
-  assert(kmap->md.generation);
 
-  if (umap->md.generation < kmap->md.generation)
-    update_kernel_pd(umap);
+  /* We're protected by unique `kernel_pmap.mtx` */
+  assert(mtx_owned(&kmap->mtx));
+
+  old_kva = roundup(old_kva, L0_SIZE);
+
+  /* Cannot allocate memory under spin lock, so do it here. */
+  for (vaddr_t va = old_kva; va <= new_kva; va += L0_SIZE)
+    kmap_l0_pages[L0_INDEX(va)] = pde_alloc(kmap);
+
+  WITH_MTX_LOCK (&kmap_l0_lock) {
+    /* Fill level 0 page directory with new pages. */
+    for (vaddr_t va = old_kva; va <= new_kva; va += L0_SIZE) {
+      pde_t *pdep = pde_ptr(kmap->pde, 0, va);
+      assert(!pde_valid_p(pdep));
+      *pdep = pde_make(0, kmap_l0_pages[L0_INDEX(va)]);
+    }
+
+    kmap->md.generation++;
+    assert(kmap->md.generation);
+  }
+}
+
+void pmap_md_activate(pmap_t *umap) {
+  update_kernel_pd(umap);
 
   __set_satp(umap->md.satp);
   __sfence_vma();
 }
 
+void pmap_md_update(pmap_t *umap) {
+  if (!umap)
+    return;
+
+  update_kernel_pd(umap);
+  __sfence_vma();
```

Comment:
If the generation is up to date, then the `__sfence_vma` is needless. You could move this statement after the `memcpy` invocation.

Reply:
Done!